src/core/expression/qgsexpressionnodeimpl.cpp

```diff
@@ -1793,6 +1793,128 @@ QList<const QgsExpressionNode *> QgsExpressionNodeInOperator::nodes() const
   return lst;
 }
 
+
+QgsExpressionNodeBetweenOperator::~QgsExpressionNodeBetweenOperator()
+{
+  delete mNode;
+  delete mLowerBound;
+  delete mHigherBound;
+}
+
+QgsExpressionNode::NodeType QgsExpressionNodeBetweenOperator::nodeType() const
+{
+  return ntBetweenOperator;
+}
+
+bool QgsExpressionNodeBetweenOperator::prepareNode( QgsExpression *parent, const QgsExpressionContext *context )
+{
+  bool res = mNode->prepare( parent, context );
+  res = res && mLowerBound->prepare( parent, context );
+  res = res && mHigherBound->prepare( parent, context );
+  return res;
+}
+
+QVariant QgsExpressionNodeBetweenOperator::evalNode( QgsExpression *parent, const QgsExpressionContext *context )
+{
+  QgsExpressionNodeBinaryOperator lowBound { QgsExpressionNodeBinaryOperator::BinaryOperator::boGE, mNode->clone(), mLowerBound->clone() };
+  QgsExpressionNodeBinaryOperator highBound { QgsExpressionNodeBinaryOperator::BinaryOperator::boLE, mNode->clone(), mHigherBound->clone() };
+  const QVariant lowBoundValue = lowBound.eval( parent, context );
+  const QVariant highBoundValue = highBound.eval( parent, context );
+  if ( lowBoundValue.isNull() || highBoundValue.isNull() )
```

Comment:
```suggestion
  const QVariant lowBoundValue = lowBound.eval( parent, context );
   if ( lowBoundValue.isNull() )
   {
     return QVariant();
   }
  const QVariant highBoundValue = highBound.eval( parent, context );
  if ( highBoundValue.isNull() )
```

avoids potentially evaluating the high bound when we can shortcut early

Reply:
actually you could also move the construction of highBound to after testing lowBoundValue too, that'll potentially save some work.

Reply:
Actually, looking at the NULL handling in the oracle implementation https://oracle.su/docs/11g/timesten.112/e13070/ttsql269.htm I think I'll need to make a smarter implementation and do not always return NULL...